pkg/daemon/vm_controller.go

```
@@ -75,80 +90,230 @@ func (r *VMReconciler) reconcile(ctx context.Context, vm *virtv1alpha1.VirtualMa
+					}
+
+					sendMigrationErrChan := make(chan error, 1)
+					go func() {
+						if err := r.getCloudHypervisorClient(vm).VmSendMigration(ctx, &cloudhypervisor.SendMigrationData{
+							DestinationUrl: "unix:/var/run/virtink/send-migration.sock",
+						}); err != nil {
+							sendMigrationErrChan <- err
+						}
+					}()
+					migrationControlBlock.SendDomainErrCh = sendMigrationErrChan
+					vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationRunning
+				}
+			case virtv1alpha1.VirtualMachineMigrationRunning:
+				if vm.Status.NodeName == r.NodeName {
+					var vmPod corev1.Pod
+					vmPodKey := types.NamespacedName{
+						Name:      vm.Status.VMPodName,
+						Namespace: vm.Namespace,
+					}
+					if err := r.Get(ctx, vmPodKey, &vmPod); err != nil {
+						return fmt.Errorf("get VM Pod: %s", err)
+					}
+
+					if vmPod.Status.Phase == corev1.PodSucceeded {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationSent
+					} else if migrationControlBlock.SendDomainErrCh == nil {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
 					} else {
-						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Resumed", "Resumed VM")
+						select {
+						case err := <-migrationControlBlock.SendDomainErrCh:
+							if err != nil {
+								r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedMigrate", "Failed to migrate VM to %s: %s", vm.Status.Migration.TargetNodeName, err)
+								vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
+							}
+						default:
+							return errors.New("VM is migrating")
+						}
+					}
+					if sendDomainCancelFunc := migrationControlBlock.SendDomainCancelFunc; sendDomainCancelFunc != nil {
+						sendDomainCancelFunc()
 					}
-				default:
-					// ignored
 				}
+			case virtv1alpha1.VirtualMachineMigrationSent:
+				if vm.Status.Migration.TargetNodeName == r.NodeName {
+					// TODO handle VmNotCreated error
```

Solve this TODO?

---

we don't know the VM is creating or it will never be created(maybe something wrong). raise the error might be a better choice now.

and now we just raise an error during VM is migrating instead of return a `ctrl.Result` with requeue. maybe we need another PR to fix all these questions.

---

We can capture certain type of errors in the main reconcile func and convert it to a requeue. But yeah, it can be another commit